pkg/platform/api/buildplanner/model/buildexpression.go

```diff
@@ -56,9 +70,68 @@ type Requirement struct {
 type VersionRequirement map[string]string
 
 type BuildExpression struct {
-	expression   map[string]interface{}
-	solveNode    *map[string]interface{}
-	requirements []Requirement
+	Let *Let
+}
+
+type Let struct {
+	Assignments []*Var
+	In          *In
+}
+
+type Var struct {
+	Name  string
+	Value *Value
+}
+
+type Value struct {
+	Ap   *Ap
+	List *[]*Value
+	Str  *string
+	Null *Null
+
+	Assignment *Var
+	Object     *[]*Var
+	Ident      *string
+}
+
+type Null struct {
+	Null string
+}
+
+type Ap struct {
+	Name      string
+	Arguments []*Value
+}
+
+type In struct {
+	FuncCall *Ap
+	Name     *string
+}
+
+// The context type is a simple stack that keeps track of the current
+// path in the AST. This is used to assist with identifiying special cases.
+type context []string
+
+func (s *context) push(str string) {
+	*s = append(*s, str)
+}
+
+func (s *context) pop() {
+	if len(*s) == 0 {
+		multilog.Error("context.pop() called on empty context stack")
+		return
+	}
+
+	*s = (*s)[:len(*s)-1]
+}
+
+func (s *context) contains(str string) bool {
+	for _, v := range *s {
+		if v == str {
+			return true
+		}
+	}
+	return false
 }
```

Comment:
This feels like some convenience helper function rather than anything to do with build expressions. I'd rather we use the native types and write out the slightly more complex push/pop logic. This feels like we're trying to replicate the API of another language inside Golang.

Another option is to use the sliceutils and strutils packages. Though I think really the only thing you might need is strutils.GetString to replace the `contains()` function. The other two IMO don't need abstraction.

Reply:
Decided to go with a new set of functions in `sliceutils`